src/NHibernate/AdoNet/ConnectionManager.cs

```diff
@@ -312,117 +300,151 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
 		{
-			log.Debug("after autocommit");
+			_log.Debug("After autocommit");
 		}
 
 		private bool IsAfterTransactionRelease
-		{
-			get { return connectionReleaseMode == ConnectionReleaseMode.AfterTransaction; }
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.AfterTransaction;
 
 		private bool IsOnCloseRelease
-		{
-			get { return connectionReleaseMode == ConnectionReleaseMode.OnClose; }
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.OnClose;
 
 		private bool IsAggressiveRelease
-		{
-			get
-			{
-				if (connectionReleaseMode == ConnectionReleaseMode.AfterTransaction)
-				{
-					return !IsInActiveTransaction;
-				}
-				return false;
-			}
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.AfterTransaction && !IsInActiveTransaction;
 
 		public ISessionFactoryImplementor Factory
-		{
-			get { return session.Factory; }
-		}
+			=> _session.Factory;
 
 		public bool IsReadyForSerialization
 		{
 			get
 			{
-				if (ownConnection)
-				{
-					return connection == null && !batcher.HasOpenResources;
-				}
-				else
+				if (_ownConnection)
 				{
-					return connection == null;
+					return _connection == null && !_batcher.HasOpenResources;
 				}
+				return _connection == null;
 			}
 		}
 
 		/// <summary> The batcher managed by this ConnectionManager. </summary>
 		public IBatcher Batcher
+			=> _batcher;
+
+		public DbCommand CreateCommand()
 		{
-			get { return batcher; }
+			var result = GetConnection().CreateCommand();
+			Transaction.Enlist(result);
+			return result;
 		}
 
+		/// <summary>
+		/// Enlist the connection into provided transaction if the connection should be enlisted.
+		/// Does nothing in case an explicit transaction ongoing.
+		/// </summary>
+		/// <param name="transaction">The transaction in which the connection should be enlisted.</param>
+		public void EnlistIfRequired(System.Transactions.Transaction transaction)
+		{
+			if (_connection == null ||
+				!ShouldAutoJoinTransaction ||
+				// Most connections do not support enlisting in a system transaction while already participating
+				// in a local transaction. They are not supposed to be mixed anyway.
+				IsInActiveExplicitTransaction ||
+				// This is not a reference equal, == operator is overridden for that type. Thus testing last.
+				transaction == _connectionAmbientTransaction)
+				return;
+
+			// Some drivers do not support enlistment with null. Skip for them, they are supposed
+			// to un-enlist by themselves.
+			if (transaction == null && !Factory.ConnectionProvider.Driver.SupportsNullEnlistment)
+			{
+				return;
+			}
+
+			// Will fail if the connection is already enlisted in another not yet completed transaction.
+			// Probable case: nested transaction scope. Supporting this could be done by releasing the
+			// connection instead of enlisting.
+			_session.Connection.EnlistTransaction(transaction);
+			_connectionAmbientTransaction = transaction;
+		}
+
+		private bool RequireConnectionSwapInDtc
```

Comment:
This swap avoids race condition between the connection own prepare phase and the session prepare phase. This swap causes two connections to participate in the transaction if the session has something to Flush. To avoid this, users will have to explicitly `Flush` the session instead of relying on scope completion for doing that for them.